ROCm specific fix to enable GPU unit tests to pass.

rsanthanam-amd opened the pull request:
The reason this is necessary is because of recent changes to the ROCm specific convrunner which requires the workspace_size attribute to be present in the algorithm descriptor.

/cc @cheshire @chsigg @jurahul

jurahul commented:
This is best review by @awpr as he is working on this code.

rsanthanam-amd commented:
For more clarity, this fix is needed because of this code:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/rocm/rocm_dnn.cc#L3199-L3203

awpr commented:
Thanks for finding this!

I don't think the code can be correct as-is: the fields `conv` and `algorithm` are part of the same oneof group, so using them both at the same time is suspicious, since one or the other must be getting a default value.  It looks like this will always set a workspace size of 0.

I think the right fix would be to make `conv_ops_gpu.cc:299` set `mutable_algorithm()` directly to the whole `AlgorithmProto` (`profile_result.algorithm().ToProto()`), and update this function to expect the full proto in `algorithm()` rather than rebuild it from its components.

rsanthanam-amd committed:
ROCm specific fix to enable GPU unit tests to pass.

The reason this is necessary is because of recent changes to the ROCm specific convrunner which requires the workspace_size attribute to be present in the algorithm descriptor.

rsanthanam-amd commented:
@awpr I have updated the PR based on your feedback.

awpr reviewed (approved):
Looks good, thanks for the fix!
